### PR TITLE
Before join operation is added to NodeExtension interface.

### PR DESCRIPTION
**Pull request by kobalski**
This interface will be implemented by EnterpriseNodeExtension and will be
used for license check of HD Memory.

### CODE REVIEW

**Comment by serkan-ozal** [comment]
:+1:

File: hazelcast/src/main/java/com/hazelcast/instance/DefaultNodeExtension.java [hunk 1]
```diff
@@ -186,4 +186,7 @@ public void destroy() {
         logger.info("Destroying node NodeExtension.");
     }
 
+    @Override
+    public void beforeJoin() { logger.info("Before join check."); }
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/instance/DefaultNodeExtension.java]: We don't do such one liners. Please use 

```
public void beforeJoin() { 
    ogger.info("Before join check."); 
}
```

**Comment by pveentjer** [comment]
I don't see any tests. Can you add them? 

Can you also add documentation. This cluster functionality is complicated and needs to have very good documentation so that we get something other people can use/maintain.

File: hazelcast/src/main/java/com/hazelcast/cluster/impl/ClusterServiceImpl.java [hunk 2]
```diff
@@ -642,6 +643,16 @@ public void handleJoinRequest(JoinRequest joinRequest, Connection connection) {
                 }
             }
 
+            if (node.isMaster()) {
+                try {
+                    node.getNodeExtension().beforeJoin();
+                } catch (Exception e) {
+                    logger.warning(e.getMessage());
+                    sendBeforeJoinCheckFailure(joinRequest.getAddress(),e.getMessage());
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/cluster/impl/ClusterServiceImpl.java]: I think you have checkstyle issues here. I guess that the cluster.impl package is fully excluded from checkstyle and that is why these issues are not shown.

File: hazelcast/src/main/java/com/hazelcast/instance/DefaultNodeExtension.java [hunk 3]
```diff
@@ -186,4 +186,9 @@ public void destroy() {
         logger.info("Destroying node NodeExtension.");
     }
 
+    @Override
+    public void beforeJoin() {
+        logger.info("Before join check.");
```
Comment (mdogan) [inline hazelcast/src/main/java/com/hazelcast/instance/DefaultNodeExtension.java]: What's the use of this log? If this method is doing nothing then no need to log anything. I think you should remove it or at least make it `FINEST`.

Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/instance/DefaultNodeExtension.java]: :+1:

**Comment by kobalski** [comment]
@pveentjer I tried to write tests for this PR, but Node,NodeEngine and ClusterServiceImpl classes are impossible to mock, especially implementation of NodeExtension is decided by NodeExtensionFactory which has a static create method. Also I could not find any tests for DefaultNodeExtension class too, for a starting point. If you have any suggestion, how to write tests for DefaultNodeExtension class, help is appreciated. Also this PR will be used for license checking for HD Cache  (https://github.com/hazelcast/hazelcast-enterprise/pull/162),so DefaultNodeExtension does not implement beforeJoin(). As for documentation, I added some java doc comments.

**Comment by pveentjer** [comment]
the NodeEngineImpl and ClusterService/PartitionServiceImpl and Node are quite difficult to deal with. Mocking probably isn't an option at all. But can't you make a test that isn't based on mocking but a true integration tests to see if you code is being triggered at all when the right conditions are met?

**Comment by kobalski** [comment]
Since beforeJoin() method is not implemented by DefaultNodeExtension on open source side, there is no way for verification whether it is called on right condition in an integration test without mocking. beforeJoin() is fully implemented in this PR (https://github.com/hazelcast/hazelcast-enterprise/pull/162/) by EnterpriseNodeExtension, so that I could trigger certain cases (eg: Cluster exceeds licensed HD cache amount) and test if it is called at right condition in an integration test. So basically, there are integration tests written in the related enterprise PR where beforeJoin() method is not empty.

**Comment by serkan-ozal** [comment]
I think, there is a way for customizing ```NodeExtension``` creation by using mocked classloader for ```Config``` object. I think, our custom classloader can behave as there is a definition for  ```NodeExtension``` in ```META-INF/services/com.hazelcast.instance.NodeExtension``` file. So this fake definition points name of our custom ```NodeExtension``` class. 

We will talk this solution with @kobalski tomorrow. However I am not really sure is this mocking stuff really needed and worth to this effort.

**Comment by kobalski** [comment]
verify

**Comment by mesutcelik** [comment]
@serkan-ozal , I don't know if that is in this PR's scope. I think we can create an enhancement issue and merge this PR if all other comments are already considered.

**Comment by serkan-ozal** [comment]
@mesutcelik right :+1: I am ok. if there is no objection, lets merge it

**Pull request merged**